security/security-spring/src/main/java/org/cbioportal/security/spring/CancerStudyPermissionEvaluator.java

```diff
@@ -262,6 +315,120 @@ public boolean hasPermission(Authentication authentication, Object targetDomainO
+    {
+        // use hashset as this list can be populated with many duplicate values
+        Set<String> studyIds = new HashSet<String>();
+        if (patientFilter.getPatientIdentifiers() != null) {
+            for (PatientIdentifier patientIdentifier : patientFilter.getPatientIdentifiers()) {
+                studyIds.add(patientIdentifier.getStudyId());
+            }
+        }
+        else {
+            uniqueKeyExtractor.extractUniqueKeys(patientFilter.getUniquePatientKeys(), studyIds);
+        }
+        return hasAccessToCancerStudies(authentication, studyIds, permission);
+    }
+
+    private boolean hasAccessToCancerStudies(Authentication authentication, SampleFilter sampleFilter, Object permission)
+    {
+        if (sampleFilter.getSampleListIds() != null) {
+            return hasAccessToSampleLists(authentication, sampleFilter.getSampleListIds(), permission);
+        }
+        else if (sampleFilter.getSampleIdentifiers() != null) {
+            return hasAccessToCancerStudiesBySampleIdentifier(authentication, sampleFilter.getSampleIdentifiers(), permission);
+        }
+        else {
+            // use hashset as this list can be populated with many duplicate values
+            Set<String> studyIds = new HashSet<String>();
+            uniqueKeyExtractor.extractUniqueKeys(sampleFilter.getUniqueSampleKeys(), studyIds);
+            return hasAccessToCancerStudies(authentication, studyIds, permission);
+        }
+    }
+
+    private boolean hasAccessToCancerStudiesBySampleIdentifier(Authentication authentication, List<SampleIdentifier> sampleIdentifiers, Object permission)
+    {
+        // use hashset as this list can be populated with many duplicate values
+        Set<String> studyIds = new HashSet<String>();
+        for (SampleIdentifier sampleIdentifier : sampleIdentifiers) {
+            studyIds.add(sampleIdentifier.getStudyId());
+        }
+        return hasAccessToCancerStudies(authentication, studyIds, permission);
+    }
+
+    private boolean hasAccessToCancerStudies(Authentication authentication, Collection<String> studyIds, Object permission)
+    {
+        for (String studyId : studyIds) {
+            if (!hasPermission(authentication, studyId, "CancerStudy", permission)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean hasAccessToMolecularProfiles(Authentication authentication, Collection<String> molecularProfileIds, Object permission)
+    {
+        List<String> profileIds = (molecularProfileIds instanceof List) ?
+            (List<String>)molecularProfileIds : new ArrayList<String>(molecularProfileIds);
+        for (MolecularProfile molecularProfile : molecularProfileRepository.getMolecularProfiles(profileIds, "SUMMARY")) {
+            if (molecularProfile == null || !hasPermission(authentication, molecularProfile, permission)) {
```

Comment:
@ersinciftci is the problem that in hasPermission `cancerStudy == null` will be true if it is a molecular profile here, and then it will look up the cancer study using the studyRepository?  Would this change fix it? `molecularProfileRepository.getMolecularProfiles(profileIds, "SUMMARY")` -> `molecularProfileRepository.getMolecularProfiles(profileIds, "DETAILED")`.  I am not sure but I *think* that the sample list will have a cancer study included and so the test `cancerStudy == null` will be false, and there will not be a call to the DB there.  It has been a long time since I have looked at this and I don't know MyBatis well, so I am not completely sure about what I am saying!

Reply:
@mandawilson You are right, if we call `DETAILED` projection for molecularProfiles and sampleLists, `cancerStudy == null` would always be false, so wouldn't make a call to db in a loop anymore.

Reply:
@ersinciftci @n1zea144 I think that for sampleLists, it already has the DETAILED projection, see here: https://github.com/cBioPortal/cbioportal/blob/rc/persistence/persistence-mybatis/src/main/java/org/cbioportal/persistence/mybatis/SampleListMyBatisRepository.java#L39  And that it is just that `molecularProfileRepository.getMolecularProfiles(profileIds, "SUMMARY")` needs to change to `molecularProfileRepository.getMolecularProfiles(profileIds, "DETAILED")`.  Though of course if the default behavior for getSampleList ever changed so that it stopped being a DETAILED projection, it would silently start accessing the DB from the loop.